File: package/node/src/server/middleware.ts
import type {
  ExpressMiddleware,
  ExpressRequest,
  ExpressResponse,
  NextFunction,
  UnauthorizedInfo,
} from "./types";
import type { ValidatedConfig } from "./config";
import { isContextMFARequired } from "./config";
import type { TokenCache } from "./cache";
import type { HookManager } from "./hooks";
import type { ServerHelpers } from "./helpers";

interface MiddlewareDependencies {
  config: ValidatedConfig;
  cache: TokenCache;
  hooks: HookManager;
  helpers: ServerHelpers;
}

class MiddlewareFactory {
  private deps: MiddlewareDependencies;

  constructor(deps: MiddlewareDependencies) {
    this.deps = deps;
  }

  authMiddleware(): ExpressMiddleware {
    return async (
      req: ExpressRequest,
      res: ExpressResponse,
      next: NextFunction,
    ): Promise<void> => {
      try {
        const token = this.extractToken(req);

        if (!token) {
          await this.triggerUnauthorized(req, "No token provided");
          res
            .status(401)
            .json({ error: "Unauthorized", message: "No token provided" });
          return;
        }

        const validation = await this.deps.helpers.validateToken(token);

        if (!validation.valid || !validation.user) {
          await this.triggerUnauthorized(req, "Invalid token", token);
          res
            .status(401)
            .json({ error: "Unauthorized", message: "Invalid token" });
          return;
        }

        req.user = validation.user;
        req.token = token;
        req.context = validation.user.context;

        next();
      } catch (err) {
        const error =
          err instanceof Error ? err.message : "Authentication failed";
        await this.triggerUnauthorized(req, error);
        res.status(401).json({ error: "Unauthorized", message: error });
      }
    };
  }

  rbacMiddleware(roles: string[]): ExpressMiddleware {
    return async (
      req: ExpressRequest,
      res: ExpressResponse,
      next: NextFunction,
    ): Promise<void> => {
      const user = req.user;

      if (!user) {
        await this.triggerUnauthorized(
          req,
          "User not authenticated",
          req.token,
        );
        res
          .status(401)
          .json({ error: "Unauthorized", message: "User not authenticated" });
        return;
      }

      const hasRole = this.deps.helpers.hasRole(user, roles);

      const hookContext = this.deps.hooks.createBaseContext(
        req.ip || "unknown",
        this.getUserAgent(req),
        req.path,
      );

      await this.deps.hooks.triggerRoleCheck({
        ...hookContext,
        user,
        requiredRoles: roles,
        hasRole,
      });

      if (!hasRole) {
        const info: UnauthorizedInfo = {
          ip: req.ip || "unknown",
          userAgent: this.getUserAgent(req),
          timestamp: Date.now(),
          reason: "Insufficient role",
          path: req.path,
          attemptedRole: roles.join(", "),
          token: req.token,
        };
        await this.deps.hooks.triggerUnauthorizedAttempt(hookContext, info);

        res.status(403).json({
          error: "Forbidden",
          message: "Insufficient permissions",
        });
        return;
      }

      next();
    };
  }

  protectRoute(roles?: string[]): ExpressMiddleware {
    return async (
      req: ExpressRequest,
      res: ExpressResponse,
      next: NextFunction,
    ): Promise<void> => {
      try {
        const token = this.extractToken(req);

        if (!token) {
          await this.triggerUnauthorized(req, "No token provided");
          res
            .status(401)
            .json({ error: "Unauthorized", message: "No token provided" });
          return;
        }

        const validation = await this.deps.helpers.validateToken(token);

        if (!validation.valid || !validation.user) {
          await this.triggerUnauthorized(req, "Invalid token", token);
          res
            .status(401)
            .json({ error: "Unauthorized", message: "Invalid token" });
          return;
        }

        req.user = validation.user;
        req.token = token;
        req.context = validation.user.context;

        if (roles && roles.length > 0) {
          const hasRole = this.deps.helpers.hasRole(validation.user, roles);

          const hookContext = this.deps.hooks.createBaseContext(
            req.ip || "unknown",
            this.getUserAgent(req),
            req.path,
          );

          await this.deps.hooks.triggerRoleCheck({
            ...hookContext,
            user: validation.user,
            requiredRoles: roles,
            hasRole,
          });

          if (!hasRole) {
            const info: UnauthorizedInfo = {
              ip: req.ip || "unknown",
              userAgent: this.getUserAgent(req),
              timestamp: Date.now(),
              reason: "Insufficient role",
              path: req.path,
              attemptedRole: roles.join(", "),
              token,
            };
            await this.deps.hooks.triggerUnauthorizedAttempt(hookContext, info);

            res.status(403).json({
              error: "Forbidden",
              message: "Insufficient permissions",
            });
            return;
          }
        }

        next();
      } catch (err) {
        const error =
          err instanceof Error ? err.message : "Authentication failed";
        await this.triggerUnauthorized(req, error);
        res.status(401).json({ error: "Unauthorized", message: error });
      }
    };
  }

  mfaMiddleware(): ExpressMiddleware {
    return async (
      req: ExpressRequest,
      res: ExpressResponse,
      next: NextFunction,
    ): Promise<void> => {
      const user = req.user;

      if (!user) {
        await this.triggerUnauthorized(
          req,
          "User not authenticated",
          req.token,
        );
        res
          .status(401)
          .json({ error: "Unauthorized", message: "User not authenticated" });
        return;
      }

      const requiresMFA = isContextMFARequired(
        user.context,
        this.deps.config.mfaRequired,
      );

      if (requiresMFA && !user.mfaVerified) {
        const hookContext = this.deps.hooks.createBaseContext(
          req.ip || "unknown",
          this.getUserAgent(req),
          req.path,
        );

        await this.deps.hooks.triggerMFARequired({
          ...hookContext,
          user,
          method: "totp",
        });

        res.status(403).json({
          error: "MFA Required",
          message: "Multi-factor authentication required",
          code: "MFA_REQUIRED",
        });
        return;
      }

      next();
    };
  }

  contextMiddleware(context: string): ExpressMiddleware {
    return async (
      req: ExpressRequest,
      res: ExpressResponse,
      next: NextFunction,
    ): Promise<void> => {
      const user = req.user;

      if (!user) {
        await this.triggerUnauthorized(
          req,
          "User not authenticated",
          req.token,
        );
        res
          .status(401)
          .json({ error: "Unauthorized", message: "User not authenticated" });
        return;
      }

      if (user.context !== context) {
        const info: UnauthorizedInfo = {
          ip: req.ip || "unknown",
          userAgent: this.getUserAgent(req),
          timestamp: Date.now(),
          reason: "Wrong context",
          path: req.path,
          token: req.token,
        };
        const hookContext = this.deps.hooks.createBaseContext(
          req.ip || "unknown",
          this.getUserAgent(req),
          req.path,
        );
        await this.deps.hooks.triggerUnauthorizedAttempt(hookContext, info);

        res.status(403).json({
          error: "Forbidden",
          message: `This endpoint requires ${context} context`,
        });
        return;
      }

      next();
    };
  }

  private extractToken(req: ExpressRequest): string | null {
    const headerValue = req.headers[this.deps.config.tokenHeader];
    return this.deps.helpers.extractTokenFromHeader(
      Array.isArray(headerValue) ? headerValue[0] : headerValue,
    );
  }

  private getUserAgent(req: ExpressRequest): string {
    const ua = req.headers["user-agent"];
    return Array.isArray(ua) ? ua[0] || "unknown" : ua || "unknown";
  }

  private async triggerUnauthorized(
    req: ExpressRequest,
    reason: string,
    token?: string,
  ): Promise<void> {
    const info: UnauthorizedInfo = {
      ip: req.ip || "unknown",
      userAgent: this.getUserAgent(req),
      timestamp: Date.now(),
      reason,
      path: req.path,
      token,
    };

    const hookContext = this.deps.hooks.createBaseContext(
      req.ip || "unknown",
      this.getUserAgent(req),
      req.path,
    );

    await this.deps.hooks.triggerUnauthorizedAttempt(hookContext, info);
  }
}

export { MiddlewareFactory };
export type { MiddlewareDependencies };

File: package/python/aether_identity/server/middleware.py
"""Aether Identity Python SDK - Server middleware."""

from typing import Optional, List, Callable, Any
from functools import wraps
from .helpers import ServerHelpers
from .hooks import HookManager
from .types import UserContext
from .config import IdentityServerConfig
from ..exceptions import AuthenticationError, AuthorizationError


class MiddlewareFactory:
    """Factory for creating middleware for various frameworks."""

    def __init__(
        self, config: IdentityServerConfig, helpers: ServerHelpers, hook_manager: HookManager
    ):
        """Initialize middleware factory.

        Args:
            config: Server configuration
            helpers: Server helpers instance
            hook_manager: Hook manager instance
        """
        self._config = config
        self._helpers = helpers
        self._hooks = hook_manager

    def _extract_token(self, request: Any) -> Optional[str]:
        """Extract token from request."""
        # Try header first
        header_name = self._config.token_header.lower()
        header_value = None

        if hasattr(request, "headers"):
            # FastAPI/Starlette style
            header_value = request.headers.get(header_name)
        elif hasattr(request, "META"):
            # Django style
            header_value = request.META.get(f"HTTP_{header_name.upper().replace('-', '_')}")
        elif hasattr(request, "headers"):
            # Flask style
            header_value = request.headers.get(header_name)

        if header_value:
            return self._helpers.extract_token_from_header(header_value)

        # Try cookie
        if hasattr(request, "cookies"):
            return request.cookies.get(self._config.cookie_name)

        return None

    def _get_client_info(self, request: Any):
        """Extract client IP and user agent."""
        ip = None
        user_agent = None

        if hasattr(request, "client") and request.client:
            ip = request.client.host
        elif hasattr(request, "META"):
            ip = request.META.get("REMOTE_ADDR")

        if hasattr(request, "headers"):
            user_agent = request.headers.get("user-agent")
        elif hasattr(request, "META"):
            user_agent = request.META.get("HTTP_USER_AGENT")

        return ip, user_agent

    # FastAPI Middleware

    def fastapi_auth_middleware(self):
        """Create FastAPI dependency for authentication."""
        from fastapi import Request, HTTPException

        async def auth_dependency(request: Request) -> UserContext:
            token = self._extract_token(request)

            if not token:
                await self._hooks.on_unauthorized_attempt(
                    reason="No token provided",
                    ip=request.client.host if request.client else None,
                    user_agent=request.headers.get("user-agent"),
                    path=str(request.url),
                )
                raise HTTPException(status_code=401, detail="Authentication required")

            user = self._helpers.validate_token(token)

            if not user:
                await self._hooks.on_unauthorized_attempt(
                    reason="Invalid token",
                    token=token,
                    ip=request.client.host if request.client else None,
                    user_agent=request.headers.get("user-agent"),
                    path=str(request.url),
                )
                raise HTTPException(status_code=401, detail="Invalid token")

            # Attach to request state
            request.state.user = user
            request.state.token = token
            request.state.context = user.context

            return user

        return auth_dependency

    def fastapi_rbac_middleware(self, roles: List[str]):
        """Create FastAPI dependency for role-based access control."""
        from fastapi import Request, HTTPException

        async def rbac_dependency(request: Request) -> UserContext:
            user = getattr(request.state, "user", None)

            if not user:
                raise HTTPException(status_code=401, detail="Authentication required")

            has_role = self._helpers.has_role(user, roles)

            await self._hooks.on_role_check(
                user_id=user.id,
                required_roles=roles,
                has_role=has_role,
                ip=request.client.host if request.client else None,
                user_agent=request.headers.get("user-agent"),
                path=str(request.url),
            )

            if not has_role:
                raise HTTPException(status_code=403, detail="Insufficient permissions")

            return user

        return rbac_dependency

    def fastapi_mfa_middleware(self):
        """Create FastAPI dependency for MFA verification."""
        from fastapi import Request, HTTPException

        async def mfa_dependency(request: Request) -> UserContext:
            user = getattr(request.state, "user", None)

            if not user:
                raise HTTPException(status_code=401, detail="Authentication required")

            if self._helpers.requires_mfa(user.context, self._config.mfa_required):
                if not user.mfa_verified:
                    await self._hooks.on_mfa_required(
                        user_id=user.id,
                        ip=request.client.host if request.client else None,
                        user_agent=request.headers.get("user-agent"),
                        path=str(request.url),
                    )
                    raise HTTPException(status_code=401, detail="MFA verification required")

            return user

        return mfa_dependency

    # Flask Middleware

    def flask_auth_middleware(self):
        """Create Flask middleware for authentication."""
        from flask import request, g

        def middleware():
            token = self._extract_token(request)

            if not token:
                self._hooks.on_unauthorized_attempt(
                    reason="No token provided",
                    ip=request.remote_addr,
                    user_agent=request.user_agent.string if request.user_agent else None,
                    path=request.path,
                )
                return None, 401

            user = self._helpers.validate_token(token)

            if not user:
                self._hooks.on_unauthorized_attempt(
                    reason="Invalid token",
                    token=token,
                    ip=request.remote_addr,
                    user_agent=request.user_agent.string if request.user_agent else None,
                    path=request.path,
                )
                return None, 401

            g.user = user
            g.token = token
            g.context = user.context

            return None, None

        return middleware

    def flask_protect_route(self, roles: Optional[List[str]] = None):
        """Create Flask decorator for protected routes."""
        from flask import g, jsonify

        def decorator(f):
            @wraps(f)
            def wrapper(*args, **kwargs):
                # Check auth
                user = getattr(g, "user", None)

                if not user:
                    return jsonify({"error": "Authentication required"}), 401

                # Check roles if specified
                if roles:
                    has_role = self._helpers.has_role(user, roles)

                    self._hooks.on_role_check(
                        user_id=user.id, required_roles=roles, has_role=has_role
                    )

                    if not has_role:
                        return jsonify({"error": "Insufficient permissions"}), 403

                return f(*args, **kwargs)

            return wrapper

        return decorator

    # Generic middleware for any framework

    def protect_route(self, roles: Optional[List[str]] = None):
        """Generic route protection decorator.

        Works with frameworks that attach user to request object.
        """

        def decorator(f):
            @wraps(f)
            def wrapper(request, *args, **kwargs):
                # Try to get user from request
                user = getattr(request, "user", None)

                if not user:
                    raise AuthenticationError("Authentication required")

                if roles and not self._helpers.has_role(user, roles):
                    raise AuthorizationError("Insufficient permissions")

                return f(request, *args, **kwargs)

            return wrapper

        return decorator
